### PR TITLE
[SPARK-6747] [SQL] Throw an AnalysisException when unsupported Java list types used in Hive UDF

### PR DESCRIPTION
**Pull request by maropu**
The current implementation can't handle List<> as a return type in Hive UDF and
throws meaningless Match Error.
We assume an UDF below;
public class UDFToListString extends UDF {
public List<String> evaluate(Object o)
{ return Arrays.asList("xxx", "yyy", "zzz"); }
}
An exception of scala.MatchError is thrown as follows when the UDF used;
scala.MatchError: interface java.util.List (of class java.lang.Class)
at org.apache.spark.sql.hive.HiveInspectors$class.javaClassToDataType(HiveInspectors.scala:174)
at org.apache.spark.sql.hive.HiveSimpleUdf.javaClassToDataType(hiveUdfs.scala:76)
at org.apache.spark.sql.hive.HiveSimpleUdf.dataType$lzycompute(hiveUdfs.scala:106)
at org.apache.spark.sql.hive.HiveSimpleUdf.dataType(hiveUdfs.scala:106)
at org.apache.spark.sql.catalyst.expressions.Alias.toAttribute(namedExpressions.scala:131)
at org.apache.spark.sql.catalyst.planning.PhysicalOperation$$anonfun$collectAliases$1.applyOrElse(patterns.scala:95)
at org.apache.spark.sql.catalyst.planning.PhysicalOperation$$anonfun$collectAliases$1.applyOrElse(patterns.scala:94)
at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:33)
at scala.collection.TraversableLike$$anonfun$collect$1.apply(TraversableLike.scala:278)
...
To make udf developers more understood, we need to throw a more suitable exception.

### CODE REVIEW

**Comment by maropu** [comment]
@marmbrus Through the discussion of #5395, I think it is hard to support java List<>  types in SparkSQL because of type erasure. ISTM that if udf developers use this type, they'd be better to use GenericUDF interfaces instead of UDF ones. So, I re-created a PR to throw a meaningful exception when this kind of types used.

Any thought?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveUDFSuite.scala [hunk 1]
```diff
@@ -133,6 +133,32 @@ class HiveUDFSuite extends QueryTest {
...
+    sql("DROP TEMPORARY FUNCTION IF EXISTS testUDFToListString")
+    TestHive.reset()
+  }
+
+  test("UDFToListInt") {
+    val testData = TestHive.sparkContext.parallelize(StringCaseClass("") :: Nil).toDF()
+    testData.registerTempTable("inputTable")
+
+    sql(s"CREATE TEMPORARY FUNCTION testUDFToListInt AS '${classOf[UDFToListInt].getName}'")
+    intercept[AnalysisException] {
```
Comment (marmbrus) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveUDFSuite.scala]: assign the result of this function to a variable and check that the message is correct.

**Comment by marmbrus** [comment]
ok to test

**Comment by marmbrus** [comment]
This looks great!  One minor comment on the tests.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveUDFSuite.scala (hunk 1 above)
Comment (maropu) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveUDFSuite.scala]: Fixed and Does it satisfy your comment?

**Comment by maropu** [comment]
@marmbrus Ok and thanks.
After this patch merged, I'll make a same patch for Map<> because it has the same issue.

**Comment by marmbrus** [comment]
Thanks!   Merging to master.

**Pull request closed without merging**